### ISSUE TITLE
Add HTTP/2 push support to Server

### ISSUE THREAD

**Issue by vbrandl** [hyperium/hyper#1586]
What is the state of HTTP/2 server push in hyper? I couldn't find anything in the documentation. Are there any plans to implement it in the future?

**Comment by seanmonstar** [comment]
First step is that the h2 dependency would need Push support: https://github.com/carllerche/h2/issues/291

**Comment by seanmonstar** [comment]
With `h2` supporting push promises, I'd like to start proposing an API to support them in hyper, initially from the **server** side:

- We'd add a `hyper::push` module to contain these new pieces.
- Add a `pusher(req: &mut http::Request) -> hyper::Result<Pusher>` that can get a `Pusher` related the request's stream, or an `Error` explaining why a `Pusher` wasn't available. Error cases could include that its an HTTP/1 message, that the client disabled server push, and the like.
- The `Pusher` would have an `async fn push_request(&mut self, req: http::Request<()>) -> hyper::Result<()>` method that tries to send the `PUSH_PROMISE` frame to the client. Error cases could include that the request isn't a legal `PUSH_PROMISE`, that the clients max concurrent streams limit has been reached, the stream has closed, and the like.
- Once a push promise has been sent to the client, the same `http::Request` shall be passed to `Service::call`, and that future spawned in a new task, to generate the response for the `PUSH_PROMISE`.

An example using this proposal looks like this:

```rust
async fn handle(mut req: Request<Body>) -> Result<Response<Body>, E> {
    match hyper::push::pusher(&mut req) {
        Ok(mut pusher) => {
            let promise = Request::builder()
                .uri("/app.js")
                .body(())
                .unwrap();
            if let Err(e) = pusher.push_request(promise).await {
                eprintln!("push failed: {}", e);
            }
        },
        Err(e) => eprintln!("http2 pusher unavailable: {}", e),
    }

    Ok(Response::new(index_page_stream()))
}
```

**Comment by djc** [comment]
@stammw would the proposed API make sense for HTTP 3?

**Comment by stammw** [comment]
I was waiting to figure out a first implementation of push streams for `quinn-h3` before
answering this, but here is what I can say just by comparing the two specs and
looking at `hyper`'s code:

`H2` and `H3` share the same semantics on push promises / push streams, except:
1. `H3 draft` states that Promises are conveyed on response streams, in any part
   of it: before header, interleaved with DATA frames...
   
   In `RFC 7540`, I can't find an explicit statement saying which stream shall be
   used to send the promise. But using the response stream is the only option that
   makes sense to me.
   
   On this line:

   ```rust
   pusher.push_request(promise).await
   ```

   I guess `Pusher` will hold a reference to the response stream so a
   `PUSH_PROMISE` frame is sent before the response itself, on the same stream ?
   
2. In `H3`, several requests can share the same promise. A first `PUSH_ID` is
   created by a request, then if a concurrent request needs to push the same
   resource, it should send a `DUPLICATE_PUSH` with the same `PUSH_ID` to avoid
   sending the content multiple times. 
   
   It seems to me that it is possible to achieve that by holding something like
   `Map<Request, PUSH_ID>` in `Service`, so `Pusher` can decide to bind a
   `.push_request(x)` call to another one that has already started for the same
   resource. Sending only a `DUPLICATE_PUSH`.
   
I have a limited knowledge of Hyper for now, so I hope this helps!

**Comment by polc** [comment]
> Once a push promise has been sent to the client, the same http::Request shall be passed to Service::call, and that future spawned in a new task, to generate the response for the PUSH_PROMISE.

Hi ! In some case it's useful to be able to push multiple responses in the same context, for example doing 1 SQL query to fetch a collection and then push each resource individually. I like the API of `h2::server::SendPushedResponse` which let the developer choose to forward the pushed request to `Service::call` or not.

**Comment by stammw** [comment]
UPDATE: `DUPLICATE_PUSH` mechanism have been removed from the HTTP/3 spec. So I guess semantics are now identical.

**Comment by eggyal** [comment]
@seanmonstar I'm happy to take a look at this, but I'm not sure about the proposed API.  In particular:

> * Add a `pusher(req: &mut http::Request) -> hyper::Result<Pusher>` that can get a `Pusher` related the request's stream, or an `Error` explaining why a `Pusher` wasn't available. Error cases could include that its an HTTP/1 message, that the client disabled server push, and the like.
> * The `Pusher` would have an `async fn push_request(&mut self, req: http::Request<()>) -> hyper::Result<()>` method that tries to send the `PUSH_PROMISE` frame to the client. Error cases could include that the request isn't a legal `PUSH_PROMISE`, that the clients max concurrent streams limit has been reached, the stream has closed, and the like.

Clients can disable (and reenable) server push at any point during the life of a connection, by sending updated settings.  Therefore this design could create a race hazard, where a `Pusher` is created (whilst the client supports server push) but the `push_request` itself fails (because the client has disabled it in the interim).

Would it not be better simply to always permit push attempts (i.e. without any `Pusher` object) and handle all failure cases, including lack of protocol support or clients that have disabled server push, as a result of that operation?

**Comment by seanmonstar** [comment]
Hm, that's a fair point. I think, however, that we'd need to consider the likelihood of different scenarios happening. I believe it would be uncommon for clients to enable and disable push promises mid-connection. Even if they do, it should be very rare for that to happen with each request.

I do think servers may want to know if pushes would even possible be accepted before starting to create `Request`s. So by checking that first, they can potentially skip looking up other stuff or creating `Request`s. Especially if it is HTTP/1...

Then again, one option is to allow "pushing" even with HTTP/1.1, and in that case, sending `103 Early Data` or `Link` headers. This could be added later without backward compatibility problems.

**Comment by AlbertMarashi** [comment]
Is this still waiting?

**Comment by mladedav** [comment]
Support has been removed from [Chrome](https://developer.chrome.com/blog/removing-push/), [Firefox](https://www.mozilla.org/en-US/firefox/132.0/releasenotes/), and even [Nginx](https://nginx.org/en/CHANGES).

I don't think `hyper` needs to support this.

**Comment by AlbertMarashi** [comment]
Should be closed then

**Comment by Catwoman08** [comment]
With so many implementations removing push support including [curl now considering removal](https://curl.se/mail/lib-2026-08/0003.html), it seems like we could close here. If anyone has strong need let us know and we can re-open.
